{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0f425270-d5bf-4677-95cb-4a6aee0e9470",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Python script to thoroughly check heterozygosity per chromosome from a VCF file\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# Path to your VCF file\n",
    "vcf_file = 'snps_clean.vcf'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f8921cb8-cdbb-4529-a1c5-ec30a9a4ae8d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'snps_clean.vcf'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3d4023b1-a31b-4832-9432-780bdddc910e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parsing the VCF file\n",
    "data = []\n",
    "with open(vcf_file, 'r') as file:\n",
    "    for line in file:\n",
    "        if not line.startswith('#'):\n",
    "            parts = line.strip().split('\\t')\n",
    "            chrom = parts[0]\n",
    "            genotype = parts[9]\n",
    "            data.append((chrom, genotype))\n",
    "\n",
    "# Creating DataFrame\n",
    "df = pd.DataFrame(data, columns=['Chromosome', 'Genotype'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "48cb1053-4afc-4eae-9e11-5fd998d9b740",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Genotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>./.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595396</th>\n",
       "      <td>MT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595397</th>\n",
       "      <td>MT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595398</th>\n",
       "      <td>MT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595399</th>\n",
       "      <td>MT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595400</th>\n",
       "      <td>MT</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>595401 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome Genotype\n",
       "0               1      0/0\n",
       "1               1      0/1\n",
       "2               1      0/0\n",
       "3               1      0/0\n",
       "4               1      ./.\n",
       "...           ...      ...\n",
       "595396         MT      0/0\n",
       "595397         MT      0/0\n",
       "595398         MT      0/0\n",
       "595399         MT      0/0\n",
       "595400         MT      0/0\n",
       "\n",
       "[595401 rows x 2 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "87fcb8f8-0544-497c-8a3d-a6d4e74119d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "            Heterozygous_Count  Total_SNPs  Heterozygosity_Percent\n",
      "Chromosome                                                        \n",
      "1                        13408       46440               28.871662\n",
      "10                        8769       29171               30.060677\n",
      "11                        7898       29040               27.196970\n",
      "12                        8206       28343               28.952475\n",
      "13                        6105       21189               28.812119\n",
      "14                        5406       18638               29.005258\n",
      "15                        4996       18058               27.666408\n",
      "16                        5336       19111               27.921093\n",
      "17                        4624       17956               25.751838\n",
      "18                        4985       16480               30.248786\n",
      "19                        3293       12994               25.342466\n",
      "2                        13480       45743               29.468990\n",
      "20                        4412       14486               30.456993\n",
      "21                        2440        8450               28.875740\n",
      "22                        2575        8972               28.700401\n",
      "3                        11231       38252               29.360556\n",
      "4                        10359       33852               30.600851\n",
      "5                        10398       34361               30.261052\n",
      "6                        11578       40180               28.815331\n",
      "7                         8787       32740               26.838729\n",
      "8                         8653       30233               28.621043\n",
      "9                         7596       26583               28.574653\n",
      "MT                           0        3286                0.000000\n",
      "X                          103       18759                0.549070\n",
      "Y                            0        2084                0.000000\n"
     ]
    }
   ],
   "source": [
    "# Function to determine if genotype is heterozygous\n",
    "def is_heterozygous(gt):\n",
    "    return gt in ['0/1', '1/0']\n",
    "\n",
    "# Applying the function and counting heterozygous genotypes\n",
    "df['Heterozygous'] = df['Genotype'].apply(is_heterozygous)\n",
    "\n",
    "# Summarizing counts by chromosome\n",
    "heterozygosity_summary = df.groupby('Chromosome')['Heterozygous'].agg(['sum', 'count'])\n",
    "heterozygosity_summary.rename(columns={'sum': 'Heterozygous_Count', 'count': 'Total_SNPs'}, inplace=True)\n",
    "heterozygosity_summary['Heterozygosity_Percent'] = (heterozygosity_summary['Heterozygous_Count'] / heterozygosity_summary['Total_SNPs']) * 100\n",
    "\n",
    "# Displaying the results\n",
    "print(heterozygosity_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7fbc6094-0559-4bb1-bb1d-7ef05c2f29e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Всего отфильтровано SNP (не '0/0', '1/1', './.'): 164638\n"
     ]
    }
   ],
   "source": [
    "#у нас есть также не стандртные обозначения генотипа, отфильтруем их\n",
    "#извлечь все генотипы, кроме стандартных гомозиготных (0/0, 1/1, ./.)\n",
    "filtered_snps = []\n",
    "\n",
    "with open(vcf_file, 'r') as file:\n",
    "    for line in file:\n",
    "        if not line.startswith('#'):\n",
    "            parts = line.strip().split('\\t')\n",
    "            genotype = parts[9]\n",
    "            if genotype not in ('0/0', '1/1', './.'):\n",
    "                filtered_snps.append(parts)\n",
    "\n",
    "# Создаем DataFrame из отфильтрованных данных\n",
    "columns = ['#CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT', 'GENOTYPE']\n",
    "filtered_df = pd.DataFrame(filtered_snps, columns=columns)\n",
    "\n",
    "# Сохраняем в отдельный файл\n",
    "#filtered_df.to_csv('non_standard_genotypes.csv', sep='\\t', index=False)\n",
    "\n",
    "# Выводим количество отфильтрованных SNP\n",
    "print(f\"Всего отфильтровано SNP (не '0/0', '1/1', './.'): {len(filtered_df)}\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6deb408a-5ee5-4c20-8005-b7d73df46495",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>GENOTYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>752721</td>\n",
       "      <td>rs3131972</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PR</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>798959</td>\n",
       "      <td>rs11240777</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PR</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>891945</td>\n",
       "      <td>rs13303106</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PR</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>894573</td>\n",
       "      <td>rs13303010</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PR</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>903104</td>\n",
       "      <td>rs6696281</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PR</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164633</th>\n",
       "      <td>X</td>\n",
       "      <td>155229100</td>\n",
       "      <td>rs731478</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PR</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164634</th>\n",
       "      <td>X</td>\n",
       "      <td>155229189</td>\n",
       "      <td>rs3093467</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PR</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164635</th>\n",
       "      <td>X</td>\n",
       "      <td>155229483</td>\n",
       "      <td>rs3093470</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PR</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164636</th>\n",
       "      <td>X</td>\n",
       "      <td>155232838</td>\n",
       "      <td>rs3093493</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PR</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164637</th>\n",
       "      <td>X</td>\n",
       "      <td>155233098</td>\n",
       "      <td>rs2037999</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PR</td>\n",
       "      <td>GT</td>\n",
       "      <td>0/1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>164638 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       #CHROM        POS          ID REF ALT QUAL FILTER INFO FORMAT GENOTYPE\n",
       "0           1     752721   rs3131972   A   G    .      .   PR     GT      0/1\n",
       "1           1     798959  rs11240777   A   G    .      .   PR     GT      0/1\n",
       "2           1     891945  rs13303106   A   G    .      .   PR     GT      0/1\n",
       "3           1     894573  rs13303010   A   G    .      .   PR     GT      0/1\n",
       "4           1     903104   rs6696281   C   T    .      .   PR     GT      0/1\n",
       "...       ...        ...         ...  ..  ..  ...    ...  ...    ...      ...\n",
       "164633      X  155229100    rs731478   C   G    .      .   PR     GT      0/1\n",
       "164634      X  155229189   rs3093467   C   T    .      .   PR     GT      0/1\n",
       "164635      X  155229483   rs3093470   C   T    .      .   PR     GT      0/1\n",
       "164636      X  155232838   rs3093493   A   C    .      .   PR     GT      0/1\n",
       "164637      X  155233098   rs2037999   C   T    .      .   PR     GT      0/1\n",
       "\n",
       "[164638 rows x 10 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45c72323-4a5d-4edf-b106-aa34beabd82c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
